catalog/app/utils/PackageUri.spec.js

```diff
@@ -0,0 +1,179 @@
+        name: 'quilt/test',
+        hash: 'abc1',
+        path: 'sub/path',
+      })
+    })
+
+    it('should throw on invalid protocol', () => {
+      expect(() =>
+        PackageUri.parse('quilt+http://bucket-name#package=quilt/test'),
+      ).toThrowError(PackageUri.PackageUriError, /unsupported protocol "quilt+http:"/)
+    })
+
+    it('should throw on missing slashes', () => {
+      expect(() =>
+        PackageUri.parse('quilt+http://bucket-name#package=quilt/test'),
+      ).toThrowError(PackageUri.PackageUriError, /missing slashes/)
+    })
+
+    it('should throw on non-root registry', () => {
+      expect(() =>
+        PackageUri.parse('quilt+http://bucket-name/sub/path#package=quilt/test'),
+      ).toThrowError(
+        PackageUri.PackageUriError,
+        /non-bucket-root registries are not supported/,
+      )
+    })
+
+    it('should throw on missing "package=" part', () => {
+      expect(() =>
+        PackageUri.parse('quilt+http://bucket-name#pacakge=quilt/test'),
+      ).toThrowError(PackageUri.PackageUriError, /missing "package=" part/)
+    })
+
+    it('should throw on empty tag', () => {
+      expect(() =>
+        PackageUri.parse('quilt+http://bucket-name#package=quilt/test:'),
+      ).toThrowError(PackageUri.PackageUriError, /"package=" part: tag must not be empty/)
+    })
+
+    it('should throw on more than one ":" in "package="', () => {
+      expect(() =>
+        PackageUri.parse('quilt+http://bucket-name#package=quilt/test:latest:sup'),
+      ).toThrowError(
+        PackageUri.PackageUriError,
+        /"package=" part may contain only one ":"/,
+      )
+    })
+
+    it('should throw on empty hash', () => {
+      expect(() =>
+        PackageUri.parse('quilt+http://bucket-name#package=quilt/test@'),
+      ).toThrowError(
+        PackageUri.PackageUriError,
+        /"package=" part: hash must not be empty/,
+      )
+    })
+
+    it('should throw on more than one "@" in "package="', () => {
+      expect(() =>
+        PackageUri.parse('quilt+http://bucket-name#package=quilt/test@abc1@cde2'),
+      ).toThrowError(
+        PackageUri.PackageUriError,
+        /"package=" part may contain only one "@"/,
+      )
+    })
+
+    it('should throw on both ":" and "@" in "package="', () => {
+      expect(() =>
+        PackageUri.parse('quilt+http://bucket-name#package=quilt/test:latest@abc1'),
+      ).toThrowError(
+        PackageUri.PackageUriError,
+        /"package=" part may either contain ":" or "@"/,
+      )
+    })
+  })
+
+  describe('stringify', () => {
+    it('should throw on missing bucket', () => {
+      expect(() => PackageUri.stringify({ name: 'quilt/test' })).toThrowError(
+        /missing "bucket"/,
+      )
+    })
+
+    it('should throw on missing name', () => {
+      expect(() => PackageUri.stringify({ bucket: 'bucket' })).toThrowError(
+        /missing "name"/,
+      )
+    })
+
+    it('should throw on both hash and tag', () => {
```

Comment:
But 
```
PackageUri.parse(`'quilt+s3://bucket-name:latest@intiew'`)
```
(has both hash and tag) is not covered. And it throws error `TypeError: Cannot read property 'replace' of null`

Reply:
i've added a test for both hash and tag and fixed that error